### PR TITLE
PrincipalCovariatesRegression and SparseKernelMethods

### PR DESCRIPTION
**Pull request by bhelfrecht**
Minor notation changes to the PCovR notebook, more substantial changes to the sparse kernel notebook. 
Highlights:
-- Notation change for the unit-variance projections from V to \tilde{T}
-- Updated sparse kernel centering, consistent with manuscript
-- Addition of kernel scaling
-- Changes to the data loading function to center and scale data and kernels, mainly for structure-based analyses (though this doesn't really affect the notebooks, which use the atom-centered CSD data+properties)
-- Cleanup of the kernel computation infrastructure to provide consistency between Gaussian and linear kernels in how structure kernels are computed

### CODE REVIEW

File: notebooks/2_PrincipalCovariatesRegression.ipynb [hunk 1]
```diff
@@ -968,7 +968,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.8"
+   "version": "3.6.9"
```
Comment (rosecers) [inline notebooks/2_PrincipalCovariatesRegression.ipynb:971]: We should omit these sorts of commits where possible.

File: notebooks/4_SparseKernelMethods.ipynb [hunk 2]
```diff
@@ -1216,7 +1254,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.8"
+   "version": "3.6.9"
```
Comment (rosecers) [inline notebooks/4_SparseKernelMethods.ipynb:1260]: Same note as above.

File: notebooks/4_SparseKernelMethods.ipynb [hunk 3]
```diff
@@ -86,12 +86,26 @@
     "In sparse kernel methods, an approximate kernel is used in place of the full kernel. This approximate kernel is typically constructed according to the [Nystr&ouml;m approximation](https://en.wikipedia.org/wiki/Low-rank_matrix_approximations#Nystr%C3%B6m_approximation) [(Williams 2001)](http://papers.nips.cc/paper/1866-using-the-nystrom-method-to-speed-up-kernel-machines.pdf),\n",
     "\n",
     "\\begin{equation}\n",
-    "    \\mathbf{K} \\approx \\mathbf{\\tilde{K}}_{NN} = \\mathbf{K}_{NM} \\mathbf{K}_{MM}^{-1} \\mathbf{K}_{NM}^T,\n",
+    "    \\mathbf{K} \\approx \\mathbf{\\hat{K}}_{NN} = \\mathbf{K}_{NM} \\mathbf{K}_{MM}^{-1} \\mathbf{K}_{NM}^T,\n",
     "\\end{equation}\n",
     "\n",
     "Here, $M$ represents a subset of the $N$ total rows/columns of the kernel matrix, i.e. the kernel between a small **active set** than is selected with subsampling method, like farthest point sampling (FPS) [(Eldar 1997)](https://doi.org/10.1109/83.623193), or a CUR decomposition [(Imbalzano2018)](https://doi.org/10.1063/1.5024611), that is discussed in the [next notebook](5_CUR.ipynb)."
```
Comment (rosecers) [inline notebooks/4_SparseKernelMethods.ipynb]:
```suggestion
    "Here, $M$ represents a subset of the $N$ total rows/columns of the kernel matrix, i.e. the kernel between a small **active set** that is selected with subsampling method, like farthest point sampling (FPS) [(Eldar 1997)](https://doi.org/10.1109/83.623193), or a CUR decomposition [(Imbalzano2018)](https://doi.org/10.1063/1.5024611), that is discussed in the [next notebook](5_CUR.ipynb)."
```

File: notebooks/4_SparseKernelMethods.ipynb [hunk 4]
```diff
@@ -86,12 +86,26 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Precomputed X_train and X_test are centered and scaled.\n",
+    "# Unscale and uncenter them to build the sparse kernels\n",
+    "# so they can be directly compared to the full kernels,\n",
+    "# which are built on the raw X data\n",
```
Comment (rosecers) [inline notebooks/4_SparseKernelMethods.ipynb]: This comment is confusing. As a reader, it's unclear if we're un-centering because we *have* to or for demonstrative purposes.

File: notebooks/4_SparseKernelMethods.ipynb [hunk 5]
```diff
@@ -186,8 +171,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "This $\\mathbf{\\Phi}_{NM}$ is also centered relative to the active set, $\\mathbf{X}$ might well be centered otherwise. We must re-center according to the distribution of $\\mathbf{X}$.  \n",
-    "We introduce the centered version, $\\mathbf{\\tilde{\\Phi}}_{NM} = \\mathbf{\\Phi}_{NM} - \\mathbf{\\bar{\\Phi}}_{M}$. If you represent each element of $\\mathbf{\\Phi}$ in its summation form\n",
+    "Just as the \"full\" kernels in [the previous notebook](3_KernelMethods.ipynb) were centered, sparse kernels must also be centered relative to the train set. The goal is to enusure that the Nyström-approximated kernel $\\mathbf{\\hat{K}}_{NN}$ is centered relative to the training set. This is achieved by centering the approximate RKHS features $\\mathbf{\\Phi}_{NM}$, and we denote the centered version of the RKHS features as $\\mathbf{\\tilde{\\Phi}}_{NM} = \\mathbf{\\Phi}_{NM} - \\mathbf{\\bar{\\Phi}}_{M}$. If we represent each element of $\\mathbf{\\Phi}$ in its summation form\n",
```
Comment (rosecers) [inline notebooks/4_SparseKernelMethods.ipynb]:
```suggestion
    "Just as the \"full\" kernels in [the previous notebook](3_KernelMethods.ipynb) were centered, sparse kernels must also be centered relative to the train set. The goal is to ensure that the Nyström-approximated kernel $\\mathbf{\\hat{K}}_{NN}$ is centered relative to the training set. This is achieved by centering the approximate RKHS features $\\mathbf{\\Phi}_{NM}$, and we denote the centered version of the RKHS features as $\\mathbf{\\tilde{\\Phi}}_{NM} = \\mathbf{\\Phi}_{NM} - \\mathbf{\\bar{\\Phi}}_{M}$. If we represent each element of $\\mathbf{\\Phi}$ in its summation form\n",
```

File: notebooks/4_SparseKernelMethods.ipynb [hunk 6]
```diff
@@ -209,7 +193,7 @@
     "\n",
```
Comment (rosecers) [inline notebooks/4_SparseKernelMethods.ipynb:196]: This previous statement "It is best..." is no longer relevant given the code below.

**Review by rosecers — CHANGES_REQUESTED** [review]

File: notebooks/2_PrincipalCovariatesRegression.ipynb (hunk 1 above)
Comment (bhelfrecht) [reply notebooks/2_PrincipalCovariatesRegression.ipynb:971]: Sure, but unless we all ensure that we all use the exact same Python version, I believe this is unavoidable, as the version "updates" are done automatically by the Jupyter notebook. In my view, there's no point in worrying about minutiae like this.

File: notebooks/4_SparseKernelMethods.ipynb (hunk 4 above)
Comment (bhelfrecht) [reply notebooks/4_SparseKernelMethods.ipynb]: In the pre-loaded data, the full kernels are computed using uncentered/unscaled X, but the centered/scaled X are loaded. Therefore, if we want to be able to compare the full kernels with the sparse kernels that are constructed in the notebook, the sparse kernels must also be built on the uncentered/unscaled X data. Otherwise the effective kernel width is different between the two. This more has to do with the particularities of the automated variable loading than any general kernel-building practice. I will try to reword this to make that more clear.

**Review by rosecers — APPROVED** [review]

File: notebooks/4_SparseKernelMethods.ipynb (hunk 6 above)
Comment (bhelfrecht) [reply notebooks/4_SparseKernelMethods.ipynb:196]: I think it is still relevant given the code in the cell below the markdown paragraph: the column mean is saved and used to center the test set kernel, just like is advised in the statement.

Comment (rosecers) [reply notebooks/4_SparseKernelMethods.ipynb:196]: Yes, but it is not being explicitly saved in a variable, as before.

Comment (bhelfrecht) [reply notebooks/4_SparseKernelMethods.ipynb:196]: It is, the variable is `K_center` (renamed from `barKM`, which I think is what it was before).

**Review by rosecers — APPROVED** [review]
